quickwit/quickwit-indexing/src/source/doc_file_reader.rs

```diff
@@ -0,0 +1,549 @@
+                .await?;
+            DocFileReader {
+                reader: SkipReader::new(stream, 0),
+                next_offset: offset as u64,
+            }
+        };
+        Ok(reader)
+    }
+
+    /// Reads the next record from the underlying file. Returns `None` when EOF
+    /// is reached.
+    pub async fn next_record(&mut self) -> anyhow::Result<Option<FileRecord>> {
+        let mut buf = String::new();
+        // TODO add retry if read stream is broken because of a transient error
+        let (bytes_read, is_last) = self.reader.read_line_and_peek(&mut buf).await?;
+        if bytes_read == 0 {
+            Ok(None)
+        } else {
+            self.next_offset += bytes_read as u64;
+            Ok(Some(FileRecord {
+                next_offset: self.next_offset,
+                doc: Bytes::from(buf),
+                is_last,
+            }))
+        }
+    }
+}
+
+pub struct ObjectUriBatchReader {
+    partition_id: PartitionId,
+    reader: DocFileReader,
+    current_offset: usize,
+    is_eof: bool,
+}
+
+impl ObjectUriBatchReader {
+    pub async fn try_new(
+        storage_resolver: &StorageResolver,
+        partition_id: PartitionId,
+        uri: &Uri,
+        position: Position,
+    ) -> anyhow::Result<Self> {
+        let current_offset = match position {
+            Position::Beginning => 0,
+            Position::Offset(offset) => offset
+                .as_usize()
+                .context("file offset should be stored as usize")?,
+            Position::Eof(_) => {
+                return Ok(ObjectUriBatchReader {
+                    partition_id,
+                    reader: DocFileReader::empty(),
+                    current_offset: 0,
+                    is_eof: true,
+                })
+            }
+        };
+        let reader = DocFileReader::from_uri(storage_resolver, uri, current_offset).await?;
+        Ok(ObjectUriBatchReader {
+            partition_id,
+            reader,
+            current_offset,
+            is_eof: false,
+        })
+    }
+
+    pub async fn read_batch(
+        &mut self,
+        source_progress: &Progress,
+        source_type: SourceType,
+    ) -> anyhow::Result<BatchBuilder> {
+        let limit_num_bytes = self.current_offset + BATCH_NUM_BYTES_LIMIT as usize;
+        let mut new_offset = self.current_offset;
+        let mut batch_builder = BatchBuilder::new(source_type);
+        while new_offset < limit_num_bytes {
+            if let Some(record) = source_progress
+                .protect_future(self.reader.next_record())
+                .await?
+            {
+                new_offset = record.next_offset as usize;
+                batch_builder.add_doc(record.doc);
+                if record.is_last {
+                    self.is_eof = true;
+                    break;
+                }
+            } else {
+                self.is_eof = true;
+                break;
+            }
+        }
+        if batch_builder.num_bytes > 0 {
```

Comment:
Is the `batch_builder.num_bytes > 0` necessary here?

Reply:
Won't this be a problem if the file is empty?

Reply:
this is what's in the current `FileSource`:
https://github.com/quickwit-oss/quickwit/blob/69f700aa72930b9a395467886ce79b8f7eeabcb6/quickwit/quickwit-indexing/src/source/file_source.rs#L95-L114

I was wondering if it was fine sending empty batches through the pipeline. I'll take a deeper look to that and remove the condition if it's OK.